File: sdk/python/wapi/_http.py
"""Transport.

Zero runtime dependencies — ``urllib.request`` from the standard library, mirroring the
TypeScript client's use of global ``fetch``. An HTTP client that drags in a dependency tree is a
liability in something meant to be dropped into other people's projects, and ``requests`` or
``httpx`` would buy little here: this is JSON in, JSON out, one header.

Synchronous on purpose. The TypeScript client is async because that is the idiom there; most
Python callers of a WhatsApp API are scripts and workers, and a sync client composes with
``asyncio.to_thread`` more easily than the reverse.
"""

from __future__ import annotations

import json
import urllib.error
import urllib.parse
import urllib.request
from typing import Any

from .errors import WapiUnavailableError, error_for

DEFAULT_BASE_URL = "https://api.wapi.crafter.run"
DEFAULT_TIMEOUT = 30.0


class Transport:
    def __init__(
        self,
        api_key: str,
        base_url: str = DEFAULT_BASE_URL,
        timeout: float = DEFAULT_TIMEOUT,
        headers: dict[str, str] | None = None,
        opener: urllib.request.OpenerDirector | None = None,
    ) -> None:
        if not api_key:
            raise ValueError("wapi: api_key is required")
        self.api_key = api_key
        # Trailing slashes would produce `//api/...`, which some proxies redirect and others 404.
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self.extra_headers = headers or {}
        # Injectable so a test can substitute a double without patching urllib globally.
        self.opener = opener or urllib.request.build_opener()

    def request(
        self,
        method: str,
        path: str,
        query: dict[str, Any] | None = None,
        body: Any = None,
        timeout: float | None = None,
    ) -> Any:
        """One request.

        Returns the *whole* body rather than unwrapping ``data``, because this API has five
        different success envelopes: ``{success, data}`` for most routes, a bare ``{status}`` for
        ``/api/status``, ``api_key`` at the top level for regenerate-key, ``publicUrl`` at the top
        level for upload and decrypt-media, and ``204`` with no body for delete. A single
        ``unwrap(res["data"])`` helper is wrong for four of those, so unwrapping is the caller's
        decision — see ``data()`` below.
        """
        url = self.base_url + path
        if query:
            pairs = {k: str(v) for k, v in query.items() if v is not None}
            if pairs:
                url = f"{url}?{urllib.parse.urlencode(pairs)}"

        payload = None if body is None else json.dumps(body).encode("utf-8")
        req = urllib.request.Request(url, data=payload, method=method)
        for key, value in self.extra_headers.items():
            req.add_header(key, value)
        # Added last so a caller cannot override the credential through `headers`.
        req.add_header("Authorization", f"Bearer {self.api_key}")
        req.add_header("Content-Type", "application/json")

        try:
            with self.opener.open(req, timeout=timeout or self.timeout) as response:
                status = response.status
                text = response.read().decode("utf-8", errors="replace")
        except urllib.error.HTTPError as exc:
            # urllib raises on 4xx/5xx, so the failure path reads the body here rather than below.
            status = exc.code
            text = exc.read().decode("utf-8", errors="replace")
            raise error_for(status, _parse(text)) from None
        except Exception as exc:  # timeouts, DNS, refused connections
            raise WapiUnavailableError(0, f"wapi: {exc}", None) from None

        # `204` carries no body; parsing it as JSON raises.
        if status == 204 or not text:
            return None
        return _parse(text)


def _parse(text: str) -> Any:
    try:
        return json.loads(text)
    except json.JSONDecodeError:
        # A non-JSON body from a proxy or gateway is still a failure worth surfacing intact.
        return text


def data(body: Any) -> Any:
    """Pull ``data`` out of the common envelope.

    Only for the routes that use it. The four that do not are handled explicitly at their call
    sites, which is deliberate: making the exception visible in the resource file beats a clever
    helper that silently returns the wrong thing.
    """
    return body["data"]

File: sdk/typescript/src/http.ts
import { errorFor, WapiUnavailableError } from "./errors.ts";

/**
 * Transport.
 *
 * Zero runtime dependencies — global `fetch`, available on Node 18+, Bun and Deno. An HTTP
 * client that pulls a dependency tree is a liability in something meant to be dropped into other
 * people's projects.
 */

export type WapiClientOptions = {
  /** Session API key or Personal Access Token. Which one depends on the endpoint. */
  apiKey: string;
  /** Defaults to the hosted deployment. */
  baseUrl?: string;
  /** Per-request deadline. The failure mode of a WhatsApp call is silence, not an error. */
  timeoutMs?: number;
  /** Swap in a custom fetch — a proxy agent, a test double, an instrumented wrapper. */
  fetch?: typeof globalThis.fetch;
  /** Extra headers on every request. Cannot override `Authorization`. */
  headers?: Record<string, string>;
};

export type RequestOptions = {
  query?: Record<string, string | number | boolean | undefined>;
  body?: unknown;
  /** Overrides the client default for one call — uploads and sends are slower than reads. */
  timeoutMs?: number;
  signal?: AbortSignal;
};

const DEFAULT_BASE_URL = "https://api.wapi.crafter.run";
const DEFAULT_TIMEOUT_MS = 30_000;

export class Transport {
  private readonly apiKey: string;
  private readonly baseUrl: string;
  private readonly timeoutMs: number;
  private readonly fetchImpl: typeof globalThis.fetch;
  private readonly extraHeaders: Record<string, string>;

  constructor(options: WapiClientOptions) {
    if (!options.apiKey) throw new Error("wapi: apiKey is required");
    this.apiKey = options.apiKey;
    // Trailing slashes would produce `//api/...`, which some proxies redirect and others 404.
    this.baseUrl = (options.baseUrl ?? DEFAULT_BASE_URL).replace(/\/+$/, "");
    this.timeoutMs = options.timeoutMs ?? DEFAULT_TIMEOUT_MS;
    this.fetchImpl = options.fetch ?? globalThis.fetch;
    this.extraHeaders = options.headers ?? {};
  }

  /**
   * One request.
   *
   * Returns the *whole* body rather than unwrapping `data`, because this API has five different
   * success envelopes: `{success, data}` for most routes, a bare `{status}` for `/api/status`,
   * `api_key` at the top level for regenerate-key, `publicUrl` at the top level for upload and
   * decrypt-media, and `204` with no body for delete. A single `unwrap(res.data)` helper is
   * wrong for four of those, so unwrapping is the caller's decision — see `data()` below.
   */
  async request<T>(method: string, path: string, options: RequestOptions = {}): Promise<T> {
    const url = new URL(this.baseUrl + path);
    for (const [key, value] of Object.entries(options.query ?? {})) {
      if (value !== undefined) url.searchParams.set(key, String(value));
    }

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), options.timeoutMs ?? this.timeoutMs);
    // A caller-supplied signal must still work alongside our deadline.
    options.signal?.addEventListener("abort", () => controller.abort(), { once: true });

    let response: Response;
    try {
      response = await this.fetchImpl(url, {
        body: options.body === undefined ? undefined : JSON.stringify(options.body),
        headers: {
          ...this.extraHeaders,
          Authorization: `Bearer ${this.apiKey}`,
          "Content-Type": "application/json",
        },
        method,
        signal: controller.signal,
      });
    } catch (cause) {
      const timedOut = controller.signal.aborted;
      throw new WapiUnavailableError(
        0,
        timedOut ? `wapi: request timed out after ${options.timeoutMs ?? this.timeoutMs}ms` : String(cause),
        null,
      );
    } finally {
      clearTimeout(timer);
    }

    // `204` carries no body; parsing it as JSON throws.
    if (response.status === 204) return undefined as T;

    const text = await response.text();
    let body: unknown = null;
    if (text) {
      try {
        body = JSON.parse(text);
      } catch {
        // A non-JSON body from a proxy or gateway is still a failure worth surfacing intact.
        body = text;
      }
    }

    if (!response.ok) throw errorFor(response.status, body);
    return body as T;
  }
}

/**
 * Pull `data` out of the common envelope.
 *
 * Only for the routes that use it. The four that do not are handled explicitly at their call
 * sites, which is deliberate: making the exception visible in the resource file is better than a
 * clever helper that silently returns the wrong thing.
 */
export const data = <T>(body: { data: T }): T => body.data;
